src/charm.py

```
@@ -1,19 +1,497 @@
+
+        Returns:
+            A string representing the hostname of the MongoDB unit.
+        """
+        unit_id = unit.name.split("/")[1]
+        return f"{self.app.name}-{unit_id}.{self.app.name}-endpoints"
+
+    # END: helper functions
+
+    # BEGIN: properties
+    @property
+    def mongos_initialised(self) -> bool:
+        """Check if mongos is initialised."""
+        return "mongos_initialised" in self.app_peer_data
+
+    @mongos_initialised.setter
+    def mongos_initialised(self, value: bool):
+        """Set the mongos_initialised flag."""
+        if value:
+            self.app_peer_data["mongos_initialised"] = str(value)
+        elif "mongos_initialised" in self.app_peer_data:
+            del self.app_peer_data["mongos_initialised"]
+
+    @property
+    def _unit_ip(self) -> str:
+        """Returns the ip address of the unit."""
+        return str(self.model.get_binding(Config.Relations.PEERS).network.bind_address)
+
+    @property
+    def is_external_client(self) -> Optional[str]:
+        """Returns the connectivity mode which mongos should use.
+
+        This is determined by checking the modes requested by the client(s).
+
+        TODO: Future PR. This should be modified to work for many clients.
+        """
+        if EXTERNAL_CONNECTIVITY_TAG not in self.app_peer_data:
+            return False
+
+        return json.loads(self.app_peer_data.get(EXTERNAL_CONNECTIVITY_TAG))
+
+    @property
+    def database(self) -> Optional[str]:
+        """Returns a mapping of databases requested by integrated clients.
+
+        TODO: Future PR. This should be modified to work for many clients.
+        """
+        if not self._peers:
+            logger.info("Peer relation not joined yet.")
+            # TODO future PR implement relation interface between host application mongos and use
+            # host application name in generation of db name.
+            return "mongos-database"
```

I think we should either raise an exception or return None

---

this function is already covered in the subsequent PR which is currently a WIP. I would prefer to not make this change in this PR 

If this request is blocking please let me know and I can make that change